coarse monotonic time jumps back

Please answer these questions before submitting your issue. Thanks!

1. What version of Rust are you using (`rustc --version`)?
    UTC Build Time:  2017-08-10 01:18:39
2. What operating system and processor architecture are you using?
    centos 6.8
3. What did you do?
4. What did you expect to see?
5. What did you see instead?
D0811 20:08:47.573029559   12094 env_linux.c:77]             Warning: insecure environment read function 'getenv' used
2017/08/11 20:13:47.927 panic_hook.rs:85: [ERROR] thread 'raftstore-1' panicked 'system time jumped back, 11295.363911996 -> 11295.363911995' at "src/util/time.rs:158"
stack backtrace:

```
   0:     0x7ff4e5e6969e - backtrace::backtrace::libunwind::trace
                        at /home/jenkins/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.2.3/src/backtrace/libunwind.rs:54
                         - backtrace::backtrace::trace<closure>
                        at /home/jenkins/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.2.3/src/backtrace/mod.rs:70
   1:     0x7ff4e5e69e53 - backtrace::capture::{{impl}}::new
                        at /home/jenkins/workspace/build_tikv_master/go/src/github.com/pingcap/tikv/target/release/build/backtrace-acbd9f8faa22306c/out/capture.rs:79
   2:     0x7ff4e575c1cc - tikv::util::panic_hook::set_exit_hook::{{closure}}
                        at src/util/panic_hook.rs:84
   3:     0x7ff4e5f0be2a - std::panicking::rust_panic_with_hook
                        at /checkout/src/libstd/panicking.rs:550
   4:     0x7ff4e5f0bcc4 - std::panicking::begin_panic<collections::string::String>
                        at /checkout/src/libstd/panicking.rs:511
   5:     0x7ff4e5f0bbf9 - std::panicking::begin_panic_fmt
                        at /checkout/src/libstd/panicking.rs:495
   6:     0x7ff4e576d698 - tikv::util::time::{{impl}}::elapsed
                        at src/util/time.rs:0
   7:     0x7ff4e576c388 - tikv::util::time::{{impl}}::elapsed
                        at src/util/time.rs:72
                         - tikv::util::time::{{impl}}::is_slow
                        at src/util/time.rs:76
   8:     0x7ff4e55f1b19 - tikv::raftstore::store::store::{{impl}}::timeout<tikv::server::transport::ServerTransport<tikv::server::transport::ServerRaftStoreRouter, tikv::server::resolve::PdStoreAddrResolver>,tikv::pd::async::client::RpcClient>
                        at /home/jenkins/workspace/build_tikv_master/go/src/github.com/pingcap/tikv/src/util/macros.rs:128
   9:     0x7ff4e54d0770 - mio::event_loop::{{impl}}::timer_process<tikv::raftstore::store::store::Store<tikv::server::transport::ServerTransport<tikv::server::transport::ServerRaftStoreRouter, tikv::server::resolve::PdStoreAddrResolver>, tikv::pd::async::client::RpcClient>>
                        at /home/jenkins/.cargo/registry/src/github.com-1ecc6299db9ec823/mio-0.5.1/src/event_loop.rs:368
                         - mio::event_loop::{{impl}}::run_once<tikv::raftstore::store::store::Store<tikv::server::transport::ServerTransport<tikv::server::transport::ServerRaftStoreRouter, tikv::server::resolve::PdStoreAddrResolver>, tikv::pd::async::client::RpcClient>>
                        at /home/jenkins/.cargo/registry/src/github.com-1ecc6299db9ec823/mio-0.5.1/src/event_loop.rs:300
  10:     0x7ff4e5576132 - mio::event_loop::{{impl}}::run<tikv::raftstore::store::store::Store<tikv::server::transport::ServerTransport<tikv::server::transport::ServerRaftStoreRouter, tikv::server::resolve::PdStoreAddrResolver>, tikv::pd::async::client::RpcClient>>
                        at /home/jenkins/.cargo/registry/src/github.com-1ecc6299db9ec823/mio-0.5.1/src/event_loop.rs:247
                         - tikv::raftstore::store::store::{{impl}}::run<tikv::server::transport::ServerTransport<tikv::server::transport::ServerRaftStoreRouter, tikv::server::resolve::PdStoreAddrResolver>,tikv::pd::async::client::RpcClient>
                        at /home/jenkins/workspace/build_tikv_master/go/src/github.com/pingcap/tikv/src/raftstore/store/store.rs:455
  11:     0x7ff4e54ae51f - tikv::server::node::{{impl}}::start_store::{{closure}}<tikv::pd::async::client::RpcClient,tikv::server::transport::ServerTransport<tikv::server::transport::ServerRaftStoreRouter, tikv::server::resolve::PdStoreAddrResolver>>
                        at /home/jenkins/workspace/build_tikv_master/go/src/github.com/pingcap/tikv/src/server/node.rs:308
                         - std::sys_common::backtrace::__rust_begin_short_backtrace<closure,()>
                        at /checkout/src/libstd/sys_common/backtrace.rs:136
  12:     0x7ff4e54b72fd - std::thread::{{impl}}::spawn::{{closure}}::{{closure}}<closure,()>
                        at /checkout/src/libstd/thread/mod.rs:364
                         - std::panic::{{impl}}::call_once<(),closure>
                        at /checkout/src/libstd/panic.rs:296
                         - std::panicking::try::do_call<std::panic::AssertUnwindSafe<closure>,()>
                        at /checkout/src/libstd/panicking.rs:454
  13:     0x7ff4e5f12faa - panic_unwind::__rust_maybe_catch_panic
                        at /checkout/src/libpanic_unwind/lib.rs:98
  14:     0x7ff4e5542886 - std::panicking::try<(),std::panic::AssertUnwindSafe<closure>>
                        at /checkout/src/libstd/panicking.rs:433
                         - std::panic::catch_unwind<std::panic::AssertUnwindSafe<closure>,()>
                        at /checkout/src/libstd/panic.rs:361
                         - std::thread::{{impl}}::spawn::{{closure}}<closure,()>
                        at /checkout/src/libstd/thread/mod.rs:363
                         - alloc::boxed::{{impl}}::call_box<(),closure>
                        at /checkout/src/liballoc/boxed.rs:648
  15:     0x7ff4e5f0a9b5 - alloc::boxed::{{impl}}::call_once<(),()>
                        at /checkout/src/liballoc/boxed.rs:658
                         - std::sys_common::thread::start_thread
                        at /checkout/src/libstd/sys_common/thread.rs:21
                         - std::sys::imp::thread::{{impl}}::new::thread_start
                        at /checkout/src/libstd/sys/unix/thread.rs:84
  16:     0x7ff4e4acdaa0 - start_thread
  17:     0x7ff4e4604aac - __clone
  18:                0x0 - <unknown>
```

--- a/src/util/time.rs
+++ b/src/util/time.rs
@@ -270,6 +270,10 @@ impl Instant {
         }
     }
 
+    // It is different from `elapsed_duration`, the resolution here is millisecond.
+    // The processors in an SMP system do not start all at exactly the same time
+    // and therefore the timer registers are typically running at an offset.
+    // Use millisecond resolution for ignoring the error.
     fn elapsed_duration_coarse(later: Timespec, earlier: Timespec) -> Duration {
         let later_ms = later.sec * MILLISECOND_PER_SECOND +
                        (later.nsec / NANOSECONDS_PER_MILLISECOND) as i64;
@@ -443,5 +447,16 @@ mod tests {
         // Add Duration.
         assert_eq!(late_raw + zero, late_raw);
         assert_eq!(late_coarse + zero, late_coarse);
+    }
+
+    #[test]
+    fn test_coarse_instant_on_smp() {
+        for i in 0..100000 {
+            let now = Instant::now_coarse();
+            if i % 100 == 0 {
+                thread::yield_now();
+            }
+            now.elapsed();
+        }
     }
 }

--- a/src/util/time.rs
+++ b/src/util/time.rs
@@ -447,6 +447,13 @@ mod tests {
         // Add Duration.
         assert_eq!(late_raw + zero, late_raw);
         assert_eq!(late_coarse + zero, late_coarse);
+
+        // PartialEq and PartialOrd
+        let ts = Timespec::new(1, 1);
+        let now1 = Instant::Monotonic(ts);
+        let now2 = Instant::MonotonicCoarse(ts);
+        assert_ne!(now1, now2);
+        assert_eq!(now1.partial_cmp(&now2), None);
     }
 
     #[test]

--- a/src/util/time.rs
+++ b/src/util/time.rs
@@ -155,6 +155,7 @@ use self::inner::monotonic_now;
 use self::inner::monotonic_coarse_now;
 
 const NANOSECONDS_PER_SECOND: u64 = 1_000_000_000;
+const MILLISECOND_PER_SECOND: i64 = 1_000;
 const NANOSECONDS_PER_MILLISECOND: i32 = 1_000_000;
 
 #[cfg(not(target_os = "linux"))]
@@ -261,8 +262,7 @@ impl Instant {
 
     fn elapsed_duration(later: Timespec, earlier: Timespec) -> Duration {
         if later >= earlier {
-            Duration::new((later.sec - earlier.sec) as u64,
-                          (later.nsec - earlier.nsec) as u32)
+            (later - earlier).to_std().unwrap()
         } else {
             panic!("system time jumped back, {:.9} -> {:.9}",
                    earlier.sec as f64 + earlier.nsec as f64 / NANOSECONDS_PER_SECOND as f64,
@@ -271,15 +271,16 @@ impl Instant {
     }
 
     fn elapsed_duration_coarse(later: Timespec, earlier: Timespec) -> Duration {
-        let sec = later.sec - earlier.sec;
-        let later_nsec = later.nsec - later.nsec % NANOSECONDS_PER_MILLISECOND;
-        let earlier_nsec = earlier.nsec - earlier.nsec % NANOSECONDS_PER_MILLISECOND;
-        if sec > 0 || (sec == 0 && later_nsec >= earlier_nsec) {
-            Duration::new(sec as u64, (later_nsec - earlier_nsec) as u32)
+        let later_ms = later.sec * MILLISECOND_PER_SECOND +
+                       (later.nsec / NANOSECONDS_PER_MILLISECOND) as i64;
+        let earlier_ms = earlier.sec * MILLISECOND_PER_SECOND +
+                         (earlier.nsec / NANOSECONDS_PER_MILLISECOND) as i64;
+        if later_ms >= earlier_ms {
+            Duration::from_millis((later_ms - earlier_ms) as u64)
         } else {
             panic!("system time jumped back, {:.3} -> {:.3}",
-                   earlier.sec as f64 + earlier_nsec as f64 / NANOSECONDS_PER_SECOND as f64,
-                   later.sec as f64 + later_nsec as f64 / NANOSECONDS_PER_SECOND as f64);
+                   earlier.sec as f64 + earlier.nsec as f64 / NANOSECONDS_PER_SECOND as f64,
+                   later.sec as f64 + later.nsec as f64 / NANOSECONDS_PER_SECOND as f64);
         }
     }
 }

--- a/src/util/time.rs
+++ b/src/util/time.rs
@@ -443,12 +443,4 @@ mod tests {
         assert_eq!(late_raw + zero, late_raw);
         assert_eq!(late_coarse + zero, late_coarse);
     }
-
-    #[test]
-    #[should_panic]
-    fn test_instant_panic() {
-        let earlier = Instant::now();
-        thread::sleep(Duration::from_millis(10));
-        Instant::now_coarse().duration_since(earlier);
-    }
 }

--- a/src/util/time.rs
+++ b/src/util/time.rs
@@ -149,24 +149,13 @@ impl Drop for Monitor {
     }
 }
 
-#[inline]
-fn elapsed_duration(later: Timespec, earlier: Timespec) -> Duration {
-    if later >= earlier {
-        Duration::new((later.sec - earlier.sec) as u64,
-                      (later.nsec - earlier.nsec) as u32)
-    } else {
-        panic!("system time jumped back, {:.9} -> {:.9}",
-               earlier.sec as f64 + earlier.nsec as f64 / NANOSECONDS_PER_SECOND as f64,
-               later.sec as f64 + later.nsec as f64 / NANOSECONDS_PER_SECOND as f64);
-    }
-}
-
 /// `monotonic_raw_now` returns the monotonic raw time since some unspecified starting point.
 pub use self::inner::monotonic_raw_now;
 use self::inner::monotonic_now;
 use self::inner::monotonic_coarse_now;
 
 const NANOSECONDS_PER_SECOND: u64 = 1_000_000_000;
+const NANOSECONDS_PER_MILLISECOND: i32 = 1_000_000;
 
 #[cfg(not(target_os = "linux"))]
 mod inner {
@@ -228,7 +217,7 @@ mod inner {
 /// A measurement of a monotonically increasing clock.
 /// It's similar and meat to replace `std::time::Instant`,
 /// for providing extra features.
-#[derive(Copy, Clone, Debug, Eq)]
+#[derive(Copy, Clone, Debug)]
 pub enum Instant {
     Monotonic(Timespec),
     MonotonicCoarse(Timespec),
@@ -247,44 +236,74 @@ impl Instant {
         match *self {
             Instant::Monotonic(t) => {
                 let now = monotonic_now();
-                elapsed_duration(now, t)
+                Instant::elapsed_duration(now, t)
             }
             Instant::MonotonicCoarse(t) => {
                 let now = monotonic_coarse_now();
-                elapsed_duration(now, t)
+                Instant::elapsed_duration_coarse(now, t)
             }
         }
     }
 
     pub fn duration_since(&self, earlier: Instant) -> Duration {
-        let later = self.get_timespec();
-        let earlier = earlier.get_timespec();
-        elapsed_duration(later, earlier)
+        match (*self, earlier) {
+            (Instant::Monotonic(later), Instant::Monotonic(earlier)) => {
+                Instant::elapsed_duration(later, earlier)
+            }
+            (Instant::MonotonicCoarse(later), Instant::MonotonicCoarse(earlier)) => {
+                Instant::elapsed_duration_coarse(later, earlier)
+            }
+            _ => {
+                panic!("duration between different types of Instants");
+            }
+        }
     }
 
-    fn get_timespec(&self) -> Timespec {
-        match *self {
-            Instant::Monotonic(t) |
-            Instant::MonotonicCoarse(t) => t,
+    fn elapsed_duration(later: Timespec, earlier: Timespec) -> Duration {
+        if later >= earlier {
+            Duration::new((later.sec - earlier.sec) as u64,
+                          (later.nsec - earlier.nsec) as u32)
+        } else {
+            panic!("system time jumped back, {:.9} -> {:.9}",
+                   earlier.sec as f64 + earlier.nsec as f64 / NANOSECONDS_PER_SECOND as f64,
+                   later.sec as f64 + later.nsec as f64 / NANOSECONDS_PER_SECOND as f64);
+        }
+    }
+
+    fn elapsed_duration_coarse(later: Timespec, earlier: Timespec) -> Duration {
+        let sec = later.sec - earlier.sec;
+        let later_nsec = later.nsec - later.nsec % NANOSECONDS_PER_MILLISECOND;
+        let earlier_nsec = earlier.nsec - earlier.nsec % NANOSECONDS_PER_MILLISECOND;
+        if sec > 0 || (sec == 0 && later_nsec >= earlier_nsec) {
+            Duration::new(sec as u64, (later_nsec - earlier_nsec) as u32)
+        } else {
+            panic!("system time jumped back, {:.3} -> {:.3}",
+                   earlier.sec as f64 + earlier_nsec as f64 / NANOSECONDS_PER_SECOND as f64,
+                   later.sec as f64 + later_nsec as f64 / NANOSECONDS_PER_SECOND as f64);
         }
     }
 }
 
 impl PartialEq for Instant {
     fn eq(&self, other: &Instant) -> bool {
-        self.get_timespec().eq(&other.get_timespec())
-    }
-}
-
-impl Ord for Instant {
-    fn cmp(&self, other: &Instant) -> Ordering {
-        self.get_timespec().cmp(&other.get_timespec())
+        match (*self, *other) {
+            (Instant::Monotonic(this), Instant::Monotonic(other)) |
+            (Instant::MonotonicCoarse(this), Instant::MonotonicCoarse(other)) => this.eq(&other),
+            _ => false,
+        }
     }
 }
 
 impl PartialOrd for Instant {
     fn partial_cmp(&self, other: &Instant) -> Option<Ordering> {
-        self.get_timespec().partial_cmp(&other.get_timespec())
+        match (*self, *other) {
+            (Instant::Monotonic(this), Instant::Monotonic(other)) |
+            (Instant::MonotonicCoarse(this), Instant::MonotonicCoarse(other)) => {
+                this.partial_cmp(&other)
+            }
+            // The Order of different types of Instants is meaningless.
+            _ => None,
+        }
     }
 }
 
@@ -423,5 +442,13 @@ mod tests {
         // Add Duration.
         assert_eq!(late_raw + zero, late_raw);
         assert_eq!(late_coarse + zero, late_coarse);
+    }
+
+    #[test]
+    #[should_panic]
+    fn test_instant_panic() {
+        let earlier = Instant::now();
+        thread::sleep(Duration::from_millis(10));
+        Instant::now_coarse().duration_since(earlier);
     }
 }

